lib/cucumber/factory/multiline_parser.rb

```diff
@@ -0,0 +1,98 @@
+module Cucumber
+  class Factory
+    class MultilineParser
+
+      STEP_PREFIX_PATTERN = /\A(with|and) these attributes:/
+      DOC_STRING_PATTERN = /^\s*(.+):(.+)$/
+      DATA_TABLE_PATTERN = /^\s*\|(.+)\|(.+)\|$/
+
+      # MultilineParser allows to use doc string or data tables for factory attributes:
+      #
+      # https://cucumber.io/docs/reference#doc-strings
+      # https://cucumber.io/docs/reference#data-tables
+      #
+      # Example usage:
+      #
+      # Given there is a user with these attributes:
+      #   """
+      #   name: Jane
+      #   locked: true
+      #   """
+      #
+      # Given there is a user with the email "test@invalid.com" and these attributes:
+      #   """
+      #   name: Jane
+      #   """
+      #
+      # Given there is a user with these attributes:
+      #   | name   | Jane |
+      #   | locked | true |
+      #
+      # Given there is a user with the email "test@invalid.com" and these attributes:
+      #   | name | Jane |
+      #
+      #
+      # Example inputs:
+      #
+      # " with these attributes:\n  name: Jane\n  locked: true\n""
+      # " and these attributes:\n  name: Jane\n"
+      # " with these attributes:\n  | name   | Jane |\n  | locked | true |\n"
+      # " and these attributes:\n  | name | Jane |\n"
+
+      def initialize(multiline_attributes)
+        @multiline_attributes = multiline_attributes
+      end
+
+      def extract_attributes
+        strip_multiline_attributes
+
```

Comment:
Here you either check the type of `multiline_attributes` (either a string, or a `DataTable`). Alternatively, we might only support the table notation, in which case the whole implementation might get a bit simpler.